server/src/main/java/org/elasticsearch/http/AbstractHttpServerTransport.java

```diff
@@ -315,6 +338,37 @@ public void incomingRequest(final HttpRequest httpRequest, final HttpChannel htt
         }
     }
 
+    void updateClientStats(final HttpRequest httpRequest, final HttpChannel httpChannel) {
+        HttpStats.ClientStats clientStats = httpChannelStats.get(System.identityHashCode(httpChannel));
+        if (clientStats != null) {
+            if (clientStats.agent == null) {
+                if (httpRequest.getHeaders().containsKey("x-elastic-product-origin")) {
```

Comment:
These checks should be `httpRequest.getHeaders().get("<whatever-header>").size() > 0` rather than a `containsKey` check, because it's possible for the header list to be an empty list

(I was able to reproduce this with a malicious `telnet x 9200` and manually constructing a "bad" HTTP request with an empty `User-Agent` header)